### PR TITLE
R4R: Service/TxSizeLimit

### LINKED ISSUE #1054: Update Params gov/MaxNum and service/TxSizeLimit
- Update param ` gov / MaxNum ` to be 1,5,2
- Add new param `service/TxSizeLimit` for the service tx. Default value 4000bytes, Range [2000, 6000] bytes and can be governed

### PR DESCRIPTION
**Pull request by MrXJC**
close #1054

### CODE REVIEW

File: app/baseapp.go [hunk 1]
```diff
@@ -704,7 +704,12 @@ func (app *BaseApp) runTx(mode RunTxMode, txBytes []byte, tx sdk.Tx) (result sdk
 		return
 	}
 
-	if err := app.Engine.GetCurrentProtocol().ValidateTx(ctx, txBytes); err != nil {
+	var msgs = tx.GetMsgs()
+	if err := validateBasicTxMsgs(msgs); err != nil {
+		return err.Result()
+	}
```
Comment (wukongcheng) [inline app/baseapp.go]: `validateBasicTxMsgs ` can be done after the tx pass the tx size check.

File: app/v0/protocol_v0.go [hunk 2]
```diff
@@ -114,16 +114,38 @@ func (p *ProtocolV0) GetVersion() uint64 {
...
 	} else {
 		panic("The subspace " + auth.DefaultParamSpace + " cannot be found!")
 	}
+	if serviceMsgMum == 0 && uint32(len(txBytes)) > txSizeLimit {
+		return sdk.ErrExceedsTxSize("the tx size exceeds the limitation")
+	}
 
-	if uint32(len(txBytes)) > txSizeLimit {
+	subspace, bool = p.paramsKeeper.GetSubspace(service.DefaultParamSpace)
+	var serviceTxSizeLimit uint64
```
Comment (wukongcheng) [inline app/v0/protocol_v0.go]: We need to get the service_txsize_limit param only if `serviceMsgMum == len(msgs)`

File: modules/service/params.go [hunk 3]
```diff
@@ -267,3 +287,10 @@ func validateArbitrationTimeLimit(v time.Duration) sdk.Error {
 	}
 	return nil
 }
+
+func validateTxSizeLimit(v uint64) sdk.Error{
+	if v < 5 {
```
Comment (wukongcheng) [inline modules/service/params.go]: - Why 5?

Comment (chengwenxi) [inline modules/service/params.go]: <5 between [2000, 6000] ?

File: modules/service/params.go [hunk 4]
```diff
@@ -191,6 +208,9 @@ func validateParams(p Params) error {
 	if err := validateArbitrationTimeLimit(p.ArbitrationTimeLimit); err != nil {
 		return err
 	}
+	if err := validateTxSizeLimit(p.TxSizeLimit); err != nil {
+		return err
+	}
```
Comment (wukongcheng) [inline modules/service/params.go]: Only check the params under `main_net`.

Comment (MrXJC) [reply modules/service/params.go]: This function have done.

File: app/v0/protocol_v0.go [hunk 5]
```diff
@@ -114,17 +114,45 @@ func (p *ProtocolV0) GetVersion() uint64 {
...
+	for _, msg := range msgs {
+		if msg.Route() == service.MsgRoute {
+			serviceMsgMum++
+		}
 	}
 
-	if uint32(len(txBytes)) > txSizeLimit {
-		return sdk.ErrExceedsTxSize("the tx size exceeds the limitation")
+	if serviceMsgMum != 0 && serviceMsgMum != len(msgs) {
+		return sdk.ErrServiceTxLimit("Can't mix the service msg with other types of msg in a transaction! ")
```
Comment (HaoyangLiu) [inline app/v0/protocol_v0.go]: Can't mix service msgs with other types of msg in one transaction!

File: app/v0/protocol_v0.go [hunk 6]
```diff
@@ -114,17 +114,45 @@ func (p *ProtocolV0) GetVersion() uint64 {
...
-	subspace, bool := p.paramsKeeper.GetSubspace(auth.DefaultParamSpace)
-	var txSizeLimit uint32
-	if bool {
-		subspace.Get(ctx, auth.TxSizeLimitKey, &txSizeLimit)
-	} else {
-		panic("The subspace " + auth.DefaultParamSpace + " cannot be found!")
+func (p *ProtocolV0) ValidateTx(ctx sdk.Context, txBytes []byte, msgs []sdk.Msg) sdk.Error {
+
+	serviceMsgMum := 0
+	for _, msg := range msgs {
```
Comment (haifengxi) [inline app/v0/protocol_v0.go]: serviceMsgMum --> serviceMsgNum

Comment (MrXJC) [reply app/v0/protocol_v0.go]: ok 😲

**Pull request merged**